### PR TITLE
[mod] package.html template: additional links (a python dict)

### PR DESCRIPTION
**Pull request by dkuku**
## What does this PR do?
Refactor package template to include custom amount of links
Rename publishedDate variable to published_date
change engines that use this variable

image
image

### CODE REVIEW

File: searx/engines/hex.py [hunk 1]
```diff
@@ -46,10 +46,9 @@ def response(resp):
                 "content": meta.get("description", ""),
                 "version": meta.get("latest_version"),
                 "maintainer": ", ".join(meta.get("maintainers", [])),
-                "publishedDate": publishedDate,
-                "tags": tags,
-                "homepage": meta.get("links", {}).get("homepage"),
-                "source_code_url": meta.get("links", {}).get("github"),
+                "published_date": published_date,
```
Comment (Bnyro) [inline searx/engines/hex.py]: Please don't change that, it won't continue to work that way. The field in the html template still uses `publishedDate` (as any template in SearXNG does).

I agree that `published_date` would be better, but that should be done in a different PR and also update the documentation.

File: searx/engines/npm.py [hunk 2]
```diff
@@ -26,15 +24,14 @@
...
-            'from': (params["pageno"] - 1) * page_size,
-            'q': query,
-            'size': page_size,
+            "from": (params["pageno"] - 1) * page_size,
+            "q": query,
+            "size": page_size,
         }
     )
-    params['url'] = search_api + args
+    params["url"] = search_api + args
```
Comment (Bnyro) [inline searx/engines/npm.py]: nit: unrelated, please remove

File: searx/engines/npm.py [hunk 3]
```diff
@@ -43,23 +40,26 @@ def response(resp):
...
                 "template": "packages.html",
                 "url": package["links"]["npm"],
                 "title": package["name"],
-                'package_name': package["name"],
+                "package_name": package["name"],
                 "content": package.get("description", ""),
                 "version": package.get("version"),
                 "maintainer": package.get("author", {}).get("name"),
-                'publishedDate': publishedDate,
+                "published_date": published_date,
```
Comment (Bnyro) [inline searx/engines/npm.py]:
```suggestion
                "publishedDate": published_date,
```

File: searx/engines/pypi.py [hunk 4]
```diff
@@ -23,44 +21,44 @@
...
             {
                 "template": "packages.html",
                 "url": url,
                 "title": title,
-                'package_name': title,
+                "package_name": title,
                 "content": content,
                 "version": version,
-                'publishedDate': parser.parse(created_at),  # type: ignore
+                "published_date": parser.parse(created_at),  # type: ignore
```
Comment (Bnyro) [inline searx/engines/pypi.py]:
```suggestion
                "publishedDate": parser.parse(created_at),  # type: ignore
```

File: searx/templates/simple/result_templates/packages.html [hunk 5]
```diff
@@ -47,17 +47,24 @@
...
       <span>{{- '' -}}
-        {%- if result.homepage -%}
-          <a href="{{ result.homepage }}" target="_blank">{{ _('Project homepage') }}</a>
-        {%- endif -%}
-        {%- if result.homepage and result.source_code_url %} | {% endif -%}
-        {%- if result.source_code_url -%}
-          <a href="{{ result.source_code_url }}" target="_blank">{{ _('Source code') }}</a>
-        {%- endif -%}
+      {% for name, link in result.links.items() %}
+        {% if not loop.first %} | {% endif %}
```
Comment (Bnyro) [inline searx/templates/simple/result_templates/packages.html]: That's a nice trick, didn't know that's possible :)

**Comment by Bnyro** [comment]
Please also update the documentation at https://github.com/searxng/searxng/blob/master/docs/dev/engines/engine_overview.rst and add the new field there too!

**Comment by dkuku** [comment]
> Please also update the documentation at https://github.com/searxng/searxng/blob/master/docs/dev/engines/engine_overview.rst and add the new field there too!

I updated the pr, it's ready for re-review

**Review by Bnyro — CHANGES_REQUESTED** [review]
I personally think it would be better to keep `source_code_url`, `documentation_url` and `homepage` normal fields (not part of the dict), because they do behave differently than normal `links`. It's a good idea to display them together, but for the developer experience, they should stay separated because one would otherwise always need to look up what specific edge cases for the `links` dictionary exist (as it's not properly documented, and doesn't follow the way other templates work).

The [GitHub engine](https://github.com/searxng/searxng/blob/58320222e2725c5e370226a1df78f9e951202e05/searx/engines/github.py#L63) would require modifications too (`source_code_url` and `homepage`) looking at the current changes. But when following the suggestion above, it won't need to be modified.

File: docs/dev/engines/engine_overview.rst [hunk 6]
```diff
@@ -556,10 +556,6 @@ type.
...
-   * - homepage
-     - :py:class:`str`
-     - the url of the project's homepage
-
-   * - source_code_url
-     - :py:class:`str`
-     - the location of the project's source code
+   * - links
+     - :py:class:`dict`
+     - a dictionary of link name to link url, e.g. {"homepage": {http://example.com"}
```
Comment (Bnyro) [inline docs/dev/engines/engine_overview.rst]:
```suggestion
     - a dictionary of link name to link url, e.g. {"homepage": "http://example.com"}
```
nit: typo

**Comment by dkuku** [comment]
After thinking about it, I think the easiest way would be to link to the documentation as homepage.
When publishing on hex the documentation is required I think, I  can't publish a package when it's not generated.

On the other hand hex does not have a standardized list of keys in the link dictionary and in the example how to publish a package there is only github so this is the only one people add in 95%. But it's not typed and can be written in different casing - some examples I printed out, sometimes there may be links to other packages on github too - when I'm wrapping a rust library so it can be used in elixir I'm adding links to it.
```python
{'Authority (GitHub)': 'https://github.com/infinitered/authority', 'Authority (Hex)': 'https://hex.pm/authority', 'GitHub': 'https://github.com/infinitered/authority_ecto'}
{'Changelog': 'https://github.com/TheFirstAvenger/ecsx_persistence_ecto/blob/master/CHANGELOG.md', 'GitHub': 'https://github.com/TheFirstAvenger/ecsx_persistence_ecto'}
{'Docs': 'https://hexdocs.pm/aws_ex_ray_ecto', 'Github': 'https://github.com/lyokato/aws_ex_ray_ecto'}
{'Docs': 'https://hexdocs.pm/couchx', 'GitHub': 'https://github.com/javierg/couchx'}
{'Documentation': 'https://hexdocs.pm/carguero_ex_audit', 'GitHub': 'https://github.com/carguero/carguero_ex_audit'}
{'GitHub': 'https://github.com/elixir-ecto/ecto'}
{'GitLab': 'https://github.com/tlux/database_yaml_config_provider'}
{'Github': 'https://github.com/sheharyarn/ecto_atomized_map'}
{'Gitlab': 'https://gitlab.com/patatoid/ecto3_mnesia'}
{'Homepage': 'https://buildkite.com/test-analytics', 'Source': 'https://github.com/buildkite/test_collector_elixir'}
{'Homepage': 'https://git.centralstandard.com/open-source/date-params'}
{'github': 'https://github.com/MorphicPro/dissolver'}
{'github': 'https://github.com/rstacruz/authsense'}
{}
```

File: searx/engines/hex.py [hunk 7]
```diff
@@ -46,11 +44,24 @@ def response(resp):
...
     return results
+
+
+def get_source_code_url(links):
+    source_url = None
+    for k, v in links.items():
+        if not source_url and k.lower() in ["source", "github", "gitlab"]:
+            source_url = v
+
+    for k, v in links.items():
```
Comment (dkuku) [inline searx/engines/hex.py]: first check all the keys then if none matches check the values

**Comment by Bnyro** [comment]
> After thinking about it, I think the easiest way would be to link to the documentation as homepage. When publishing on hex the documentation is required I think, I can't publish a package when it's not generated.
> 
> On the other hand hex does not have a standardized list of keys in the link dictionary and in the example how to publish a package there is only github so this is the only one people add in 95%. But it's not typed and can be written in different casing - some examples I printed out, sometimes there may be links to other packages on github too - when I'm wrapping a rust library so it can be used in elixir I'm adding links to it.
> 
> ```python
> {'Authority (GitHub)': 'https://github.com/infinitered/authority', 'Authority (Hex)': 'https://hex.pm/authority', 'GitHub': 'https://github.com/infinitered/authority_ecto'}
> {'Changelog': 'https://github.com/TheFirstAvenger/ecsx_persistence_ecto/blob/master/CHANGELOG.md', 'GitHub': 'https://github.com/TheFirstAvenger/ecsx_persistence_ecto'}
> {'Docs': 'https://hexdocs.pm/aws_ex_ray_ecto', 'Github': 'https://github.com/lyokato/aws_ex_ray_ecto'}
> {'Docs': 'https://hexdocs.pm/couchx', 'GitHub': 'https://github.com/javierg/couchx'}
> {'Documentation': 'https://hexdocs.pm/carguero_ex_audit', 'GitHub': 'https://github.com/carguero/carguero_ex_audit'}
> {'GitHub': 'https://github.com/elixir-ecto/ecto'}
> {'GitLab': 'https://github.com/tlux/database_yaml_config_provider'}
> {'Github': 'https://github.com/sheharyarn/ecto_atomized_map'}
> {'Gitlab': 'https://gitlab.com/patatoid/ecto3_mnesia'}
> {'Homepage': 'https://buildkite.com/test-analytics', 'Source': 'https://github.com/buildkite/test_collector_elixir'}
> {'Homepage': 'https://git.centralstandard.com/open-source/date-params'}
> {'github': 'https://github.com/MorphicPro/dissolver'}
> {'github': 'https://github.com/rstacruz/authsense'}
> {}
> ```
>
I think that the `links` solution you did before was a good solution which can be used for all package links at the `hex` engine in my opinion - it's probably better than hardcoding the list of git sources (`github`, `gitlab`, ...)

I just didn't like the way you handled the translation of the strings for `source_code_url` and `homepage` before.

So my take would be - take the commit https://github.com/searxng/searxng/pull/3464/commits/8a400c896937b876196d6fbfbff546c8d402af75 as base and just make `links` work independent from `source_code_url` and `homepage` - so that all of these fields can be used as needed by the engines using the template.

**Comment by dkuku** [comment]
> So my take would be - take the commit [8a400c8](https://github.com/searxng/searxng/commit/8a400c896937b876196d6fbfbff546c8d402af75) as base and just make `links` work independent from `source_code_url` and `homepage` - so that all of these fields can be used as needed by the engines using the template.

The translations could be moved to the engine itself instead of template - then the for loop can take 5 lines.
These are engine specific, for example npm includes `bugs` link that is hidden currently.

I started to generate the links in a loop because currently it's using multiple if statements to insert the `|` divider.  [It's already complicated with 2 links (13 lines).](https://github.com/searxng/searxng/blob/383d8735976466e91c69d3d7033fc352b100d50a/searx/templates/simple/result_templates/packages.html#L50) 
I wanted to add an optional third link and it became too complicated, so I started to refactor it.

Keeping the existing url's and adding separate logic with for loop would still be a bit ugly and take around 20 lines.

**Comment by Bnyro** [comment]
> The translations could be moved to the engine itself instead of template - then the for loop can take 5 lines.
>
I think that source code and homepage are available for almost all package engines at SearXNG, so the translation should not be done by each engine individually to reduce boilerplate code.

**Comment by Bnyro** [comment]
Looks good to me now from the idea, I'm just thinking about whether it would make sense to move the `license_url` and `license_name` fields to `links` too? What's your opinion?

**Comment by dkuku** [comment]
> Looks good to me now from the idea, I'm just thinking about whether it would make sense to move the `license_url` and `license_name` fields to `links` too? What's your opinion?

In hex there is only a list of licenses available so I can't do it in this particular example - It's like tags. 
I'm not sure if linking to the actual license text makes sense at all, it may bloat the results a bit. If you are interested in the license then you probably know which one you are interested in.

**Comment by Bnyro** [comment]
> In hex there is only a list of licenses available so I can't do it in this particular example - It's like tags.
>
Yes, I see, I was rather thinking about the template for packages in general (GitHub, NPM, ...)

**Comment by return42** [comment]
@dkuku thanks a lot for your work on this topic :+1: .. I'm going to review this PR and https://github.com/searxng/searxng/pull/3487 .. both work on packages.html, so I will start with this PR .. right? .. https://github.com/searxng/searxng/pull/3487 is approved by @Bnyro so I assume this modification of packages.html is accepted by both of you in sense of https://github.com/searxng/searxng/issues/3456 .. which can be closed when this PR (#3464) is merged / right?

In the course of the changes to the PR here, the target, the package.html was taken out in the meantime and the patch history as well as the title of this PR are unclear regarding this target ... 

@dkuku is it okay for you if I clean up the commit history and also correct the title of this PR again? (if I have interpreted the course of the PR correctly with my assumptions above).

**Comment by dkuku** [comment]
@return42  👍🏻 edit both pr-s it too your liking and merge it when done.

File: docs/dev/engines/engine_overview.rst [hunk 8]
```diff
@@ -563,3 +563,7 @@ type.
    * - source_code_url
      - :py:class:`str`
      - the location of the project's source code
+
+   * - links
+     - :py:class:`dict`
+     - additional links in the form of ``{'link_name': 'http://example.com'}``
```
Comment (return42) [inline docs/dev/engines/engine_overview.rst:569]: I fixed it to rendered in a ``monospace font``

File: searx/templates/simple/result_templates/packages.html [hunk 9]
```diff
@@ -58,6 +58,14 @@
         {%- if result.source_code_url -%}
           <a href="{{ result.source_code_url }}" target="_blank">{{ _('Source code') }}</a>
         {%- endif -%}
+        {%- if result.links %}
+          {%- for name, link in result.links.items() -%}
+            {% if not loop.first or result.homepage or result.source_code_url %} | {% endif %}
+            <a href="{{ link }}" target="_blank">
+              {{- _(name) -}}
+            </a>
+          {%- endfor -%}
```
Comment (return42) [inline searx/templates/simple/result_templates/packages.html:67]: I added some `{%-` & `-%}` to get rid of not needed whitespaces in the output.

File: searx/templates/simple/result_templates/packages.html [hunk 10]
```diff
@@ -58,6 +58,14 @@
         {%- if result.source_code_url -%}
           <a href="{{ result.source_code_url }}" target="_blank">{{ _('Source code') }}</a>
         {%- endif -%}
+        {%- if result.links %}
+          {%- for name, link in result.links.items() -%}
+            {% if not loop.first or result.homepage or result.source_code_url %} | {% endif %}
+            <a href="{{ link }}" target="_blank">
+              {{- _(name) -}}
```
Comment (return42) [inline searx/templates/simple/result_templates/packages.html:65]: argueable .. but if we have a translation, then lets translate. Common words like `author` are already in our translation memory ..

EN ("Author"):

image


DE ("Autor"):  
image

Common terms can be added to [searx/searxng.msg](https://github.com/searxng/searxng/blob/master/searx/searxng.msg) where the translation of `author` also comes from (a term that is frequently used in social media):

https://github.com/searxng/searxng/blob/645a840d8211a016d808cf7da1ab62ff12d1db56/searx/searxng.msg#L87-L97

**Review by return42 — APPROVED** [review]
@dkuku @Bnyro I squashed the commit history into one commit and slightly changed the commit message.  And I did some small modifications / see [diff](https://github.com/searxng/searxng/compare/bb1d3e3b6fc54c1515909877a4b31344855bcbb6..7dc6a368b69dfe766d32542d1714e31d8dcc1a95#diff-60332926fe56944d856be975ba056cddb5ebeaf7370dc02df2bfdb17e11ad480R569)

See my comments in this review below for more details .. if you are OK with what I have done, give a short feedback to merge this PR.

Again, thanks a lot for your work on this PR :+1:

File: searx/templates/simple/result_templates/packages.html (hunk 9 above)
Comment (dkuku) [reply searx/templates/simple/result_templates/packages.html:67]: Should this name be normalised: upper or lower?

**Comment by dkuku** [comment]
Looks good to me.

**Review by Bnyro — APPROVED** [review]

File: searx/templates/simple/result_templates/packages.html (hunk 9 above)
Comment (return42) [reply searx/templates/simple/result_templates/packages.html:67]: Good question .. I hesitate to normalize in the templates .. upper-/lowercase depends on the l10n .. we have languages that do not follow the rules of the alphabet (chinese, hebraic, arabic ..) .. in my opinion, a template should be neutral with regard to i18n.

Normalization should therefore better be done in the engine ... if hex does not return a normalized value (if some results contain “Author” and others “author”) we would have to normalize this in the hex engine ... I pushed https://github.com/searxng/searxng/pull/3464/commits/5269dbdd053cc57c5cb40e4a606a7c2351730687 on top of this branch.

Is it OK to merge this commit with your PR here?

Comment (dkuku) [reply searx/templates/simple/result_templates/packages.html:67]: it's ok.
